app/chat.hoon

```diff
@@ -0,0 +1,606 @@
+    ==
+    ::  end of branching on sto type
+  ==
+  ::  end of i.wir branching
+::
+::  +lient arms
+::
+::
+::  +bound: lient tells us we successfully bound our server to the ~chat url
+::
+++  bound
+  |=  [wir=wire success=? binding=binding:http-server]
+  ^-  (quip move _this)
+  [~ this]
+::
+::  +poke-handle-http-request: serve pages from file system based on URl path
+::
+++  poke-handle-http-request
+  %-  (require-authorization:app ost.bol move this)
+  |=  =inbound-request:http-server
+  ^-  (quip move _this)
+  ::
+  =+  request-line=(parse-request-line url.request.inbound-request)
+  =/  name=@t
+    =+  back-path=(flop site.request-line)
+    ?~  back-path
+      ''
+    i.back-path
+  ?:  =(name 'tile')
+    [[ost.bol %http-response (js-response:app tile-js)]~ this]
+  ?+  site.request-line
+    :_  this
+    [ost.bol %http-response not-found:app]~
+  ::
+  ::  styling
+  ::
+      [%'~chat' %css %index ~]
+    :_  this
+    [ost.bol %http-response (css-response:app style)]~
+  ::
+  ::  javascript
+  ::
+      [%'~chat' %js %index ~]
+    :_  this
+    [ost.bol %http-response (js-response:app script)]~
+  ::
+  ::  images
+  ::
+      [%'~chat' %img *]
+    =/  img  (as-octs:mimes:html (~(got by chat-png) `@ta`name))
+    :_  this
+    [ost.bol %http-response (png-response:app img)]~
+  ::
+  ::  paginated message data
+  ::
+      [%'~chat' %scroll @t @t @t @t ~]
+    =/  cir/circle:hall  [(slav %p &3:site.request-line) &4:site.request-line]
+    =/  start/@ud  (need (rush &5:site.request-line dem))
+    =/  parsedend/@ud  (need (rush &6:site.request-line dem))
+    =*  messages  messages.str.sta
+    =/  envs/(unit (list envelope:hall))  (~(get by messages) cir)
+    ?~  envs
+      [~ this]
+    ?:  (gte start (lent u.envs))
+      [~ this]
+    =/  end/@
+      ?:  (gte parsedend (lent u.envs))
+        (dec (lent u.envs))
+      parsedend
+    =/  offset  (sub end start)
+    =/  jon/json  %-  msg-to-json
+    :*  %messages
+        cir
+        start
+        end
+        (swag [start offset] u.envs)
+    ==
+    :_  this
+    [ost.bol %http-response (json-response:app (json-to-octs jon))]~
+  ::
+  ::
+  ::  inbox page
+  ::
+     [%'~chat' *]
+    :_  this
+    [ost.bol %http-response (html-response:app index)]~
+  ==
+::
+::
+::  +subscription-retry arms
+::
+::
+::  +reap: recieve acknowledgement for peer, retry on failure
+::
+++  reap
+  |=  [wir=wire err=(unit tang)]
+  ^-  (quip move _this)
+  ?~  err
+    [~ this]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  i.wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ::
+      %circle
+      =/  shp/@p  (slav %p &2:wir)
+      =/  pat  /circle/[&3:wir]/config/group
+      ?:  =(&3:wir 'inbox')
+        :_  this
+        [ost.bol %peer wir [shp %hall] pat]~
+      ?:  (~(has in src.inbox.str.sta) [[shp &3:wir] ~])
+        :_  this
+        [ost.bol %peer wir [shp %hall] pat]~
+      [~ this]
```

Comment:
This entire expression (after `%circle`) should be de-dented.

Reply:
fixed